src/msg/async/AsyncMessenger.cc

```diff
@@ -350,6 +353,72 @@ void WorkerPool::start()
   }
 }
 
+Worker* WorkerPool::get_worker()
+{
+  ldout(cct, 10) << __func__ << dendl;
+
+   // start with some reasonably large number
+  unsigned min_load = 0x7fffffff;
+  Worker* current_best = nullptr;
+
+  simple_spin_lock(&pool_spin);
+ // find worker with least references
+  // tempting case is returning on references == 0, but in reality
+  // this will happen so rarely that there's no need for special case.
+  for (auto p = workers.begin(); p != workers.end(); ++p) {
+    unsigned worker_load = (*p)->references.read();
+    if (worker_load < min_load) {
+      current_best = *p;
+      min_load = worker_load;
+    }
+  }
+
+  // if minimum load exceeds amount of workers, make a new worker
+  // logic behind this is that we're not going to create new worker
+  // just because others have *some* load, we'll defer worker creation
+  // until others have *plenty* of load. This will cause new worker
+  // to get assigned to all new connections *unless* one or more
+  // of workers get their load reduced - in that case, this worker
+  // will be assigned to new connection.
+  // TODO: add more logic and heuristics, so connections known to be
+  // of light workload (heartbeat service, etc.) won't overshadow
+  // heavy workload (clients, etc).
+  if (!current_best || ((workers.size() < (unsigned)cct->_conf->ms_async_max_op_threads)
+      && (min_load > workers.size()))) {
+     ldout(cct, 10) << __func__ << " creating worker" << dendl;
+     current_best = new Worker(cct, this, workers.size());
+     workers.push_back(current_best);
+     current_best->create("ms_async_worker");
+  }
+
+  current_best->references.inc();
+  simple_spin_unlock(&pool_spin);
+
+  return current_best;
+}
+
+void WorkerPool::release_worker(EventCenter* c)
+{
+  ldout(cct, 10) << __func__ << dendl;
+  simple_spin_lock(&pool_spin);
+  for (auto p = workers.begin(); p != workers.end(); ++p) {
+    if (&((*p)->center) == c) {
+      ldout(cct, 10) << __func__ << " found worker, releasing" << dendl;
+      (*p)->references.dec();
+      // prevent wtf conditons, in this case we can just reset the counter
+      // worst case scenario, the workload won't be balanced.
+      unsigned oldref = (*p)->references.read();
+      if (oldref >= 0x7fffffff) {
```

Comment:
That would happen when reference count will be decreased more than once per AsyncConnection-Worker pair - this should be done only in AsyncConnection dtor, but if for some reason it would be done somewhere else too, then we'll catch it here. I don't say that it'll happen with current code, I'm just defending it against possible future bugs.